boskos/ranch/storage.go

```diff
@@ -132,67 +219,278 @@ func (s *Storage) SyncResources(data []common.Resource) error {
 	var finalError error
 
-	// delete non-exist resource
-	valid := 0
-	for _, res := range resources {
+	for _, r := range resToDelete {
 		// If currently busy, yield deletion to later cycles.
-		if res.Owner != "" {
-			resources[valid] = res
-			valid++
-			continue
-		}
-		toDelete := true
-		for _, newRes := range data {
-			if res.Name == newRes.Name {
-				resources[valid] = res
-				valid++
-				toDelete = false
-				break
+		if !r.IsInUse() {
+			logrus.Infof("Deleting resource %s", r.Name)
+			if err := s.DeleteResource(r.Name); err != nil {
+				finalError = multierror.Append(finalError, err)
+				logrus.WithError(err).Errorf("unable to delete resource %s", r.Name)
 			}
 		}
-		if toDelete {
-			logrus.Infof("Deleting resource %s", res.Name)
-			if err := s.DeleteResource(res.Name); err != nil {
+	}
+
+	for _, r := range resToAdd {
+		logrus.Infof("Adding resource %s", r.Name)
+		r.LastUpdate = s.now()
+		if err := s.AddResource(r); err != nil {
+			finalError = multierror.Append(finalError, err)
+			logrus.WithError(err).Errorf("unable to delete resource %s", r.Name)
+		}
+	}
+
+	for _, r := range resToUpdate {
+		if !r.IsInUse() {
+			logrus.Infof("Updating resource %s", r.Name)
+			if _, err := s.UpdateResource(r); err != nil {
 				finalError = multierror.Append(finalError, err)
-				logrus.WithError(err).Errorf("unable to delete resource %s", res.Name)
+				logrus.WithError(err).Errorf("unable to updateresource %s", r.Name)
 			}
 		}
 	}
-	resources = resources[:valid]
+	return finalError
+}
+
+func (s *Storage) persistDynamicResourceLifeCycles(dRLCToUpdate, dRLCToAdd, dRLCToDelelete []common.DynamicResourceLifeCycle) error {
+	var finalError error
+
+	for _, dRLC := range dRLCToDelelete {
+		logrus.Infof("Deleting resource type life cycle %s", dRLC.Type)
+		if err := s.DeleteDynamicResourceLifeCycle(dRLC.Type); err != nil {
+			finalError = multierror.Append(finalError, err)
+			logrus.WithError(err).Errorf("unable to delete resource type life cycle %s", dRLC.Type)
+		}
+	}
+
+	for _, DRLC := range dRLCToAdd {
+		logrus.Infof("Adding resource type life cycle %s", DRLC.Type)
+		if err := s.AddDynamicResourceLifeCycle(DRLC); err != nil {
+			finalError = multierror.Append(finalError, err)
+			logrus.WithError(err).Errorf("unable to add resource type life cycle %s", DRLC.Type)
+		}
+	}
+
+	for _, dRLC := range dRLCToUpdate {
+		logrus.Infof("Updating resource type life cycle %s", dRLC.Type)
+		if _, err := s.UpdateDynamicResourceLifeCycle(dRLC); err != nil {
+			finalError = multierror.Append(finalError, err)
+			logrus.WithError(err).Errorf("unable to update resource type life cycle %s", dRLC.Type)
+		}
+	}
+
+	return finalError
+}
+
+func (s *Storage) syncStaticResources(newConfig, existingResources []common.Resource) error {
+	var resToUpdate, resToAdd, resToDelete []common.Resource
+
+	// delete non-exist resource
+	newResSet := map[string]bool{}
+	remainingSet := map[string]bool{}
```

Comment:
(ditto, use actual set)

Reply:
I think using a set is overkill for that usecase.